README.md
secret
======
[![Build Status](https://travis-ci.org/TelekomCloud/puppet-secret.png)](https://travis-ci.org/TelekomCloud/puppet-secret)

Puppet function to create and manage secrets for nodes on puppet master.

    $secret_path = secret('my_id')

You now have a pointer to your secret, e.g. `puppet:///secrets/cryptsetup_key`

You can use it via:

    file { 'secretfile':
      path    => '/my/secretfile',
      ensure  => present,
      mode    => 0600,
      source  => $secret_path,
    }

Secrets that don't exist will be generated for you.

If you follow the installation correctly, you have an easy ACL-secured per-node secret store on the puppet master. As an added bonus, you can even avoid secrets from getting cached on your puppet nodes.


Installation
------------

1. Clone from git:

        git clone <source> puppet-secret

2. Install to puppet

        mv puppet-secret /etc/puppet/modules/secret

3. Create a folder on puppet master where you want to store your secrets. Ideally, this should be outside of puppet configuration (and securely replicated ;) ).

        mkdir -p /secrets

    make sure puppet has access to this folder:

        chown -R puppet:puppet /secrets

4. Configure this path in your `/etc/puppet/fileserver.conf`. Make sure it is on a per-node basis! (to do this, include the `%H` in the path as shown; this makes sure the node's FQDN is part of the access path)

        [secrets]
        path /secrets/%H/
        allow *.mydomain

5. Use it. See examples.


Options
-------

* `secrets_mount` := the mount path in fileserver, where secrets are stored (default: `secrets`)
* `bytes` := the number of bytes to use for generating a new secret (default: `128`)
* `length` := alternative to `bytes`: specifies how long the generated secret should be (number of fields).  
  for `base64`, `y64` and `alphabet` it is the number of characters in the output (without padding).
* `method` := can be either:
  * `default` := create a binary secret, i.e. select randomly from all available bits
  * `base64` := base64 encoded secret, e.g. `unM/BV7h7P6Nog==`
  * `y64` := base64 encoded secret which is url-safe, e.g. `-biHariH.ovzvw`
    (replaces ` + / = ` with ` . _ - `) (see [y64 explained](http://www.yuiblog.com/blog/2010/07/06/in-the-yui-3-gallery-base64-and-y64-encoding/ "In the YUI 3 Gallery: Base64 and Y64 encoding"))
  * `alphabet` := generate secret from the `[a-zA-Z]` letterspace, e.g. `enbpaAyuFfYSHKx`
  * `uuid` := generate UUID-based secret, e.g. `9b064a5c-0b2b-4162-b938-a43704c77602`
  * `ceph` := generate secret for ceph via `ceph-authtool`. For this, the tool must exist on the server. On ubuntu you get it by installing `ceph-common`.

more examples:

* for base64-encoded secret in 200 bytes

        $secret_path = secret('myid', {
          'bytes' => 200,
          'method' => 'base64'
          })

    which will get you something like:

        XgtLsQtcm6Tnxxpxzpo02C3geXKVo1uMMZXbohXWZWLQ3wqMrjEyTEGjImvU4/FIeXj01C+KM8R2oBu28qlLzzZX+4eaWny9n+76bRURbbZmOU7pNks5wB5lw3Y32kVlBiiiu0hMDYjqIuZ7kcwPSpO6a+Cxr/b5iToii13Ni29DXjYZq1SyPwfW3a2/qbIY4ziX3VLCRbWkzugecUVJ8mFXVniUG7Ssvu79XxXKfJJ9Vx9HbMYQJs7VAz0ZHND9FdqMknDEaIw=

* for alphabet-based secret in 16 characters

        $secret_path = secret('myid', {
          'length' => 16,
          'method' => 'alphabet'
          })

    which will get you something like:

        bnslTjPYSTHvakcQ


Additional information
======================

Limitations
-----------

Don't mistake this for a key-manager. It isn't. This is simply meant for all those task that quickly and easily need an unmanaged secret without extra components and complications.


URI-mapping
-----------

Puppet's fileserver configuration makes sure you have a node-based ACL. It is easy to configure.

With a server configuration like this:

    [secrets]
    path /secrets/%H/
    allow *.mydomain

you will create a `puppet:///secrets/` URI.

If a client called `puppy.com` accesses a secret, e.g.:

    $secret_path = secret('myid')

this will be mapped into the puppet URI:

    puppet:///secrets/myid

and on the server's filesystem will be turned into reading

    /secrets/puppy.com/myid

Through the generator, the server will make sure the requested file actually exists.
If not, the server will create it with a new secret inside.


Shared secret
-------------

This is easily possible by removing node-based ACL from puppet:

    [shared-secrets]
    path /secrets/shared/
    allow *.mydomain


License and Author
==================

Author:: Dominik Richter <do.richter@telekom.de>  
Copyright:: 2013, Deutsche Telekom AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



CHANGELOG
0.8: added uuid secret provider
0.7: added examples for shared secret
0.6: added ceph secret provider
0.5: integrated into travis-ci
0.4: select secret length via either 'bytes' or 'length'
     added rspec tests for secret
0.3: select secret-generator via 'method' = 'default/base64/y64/alphabet'
     added alphabet-based ([a-zA-Z]) secrets
0.2: added url-safe base64 encoded secrets
0.1: initial release


spec/classes/secret_spec.rb
require 'puppet'
require 'minitest/autorun'
require_relative '../../lib/puppet/parser/functions/secret'

RSpec.configure do |config|
  config.mock_framework = :mocha
end

describe Secret do
  describe '.generate_secret' do

    before do
      @test_chars = lambda{|res,alph| res.split('').find_all{|c| alph.include?(c)}.length == res.length}
      @base64 =   "+/0123456789=ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
      @y64 =      "._0123456789-ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
      @alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    end

    it 'should have a default of binary with 128 bytes' do
      res = Secret::generate_secret
      res.length.should eq(128)
    end

    it 'should generate random binary' do
      Secret::generate_secret({'method'=>'default','bytes'=>200}).length.should eq(200)
    end

    it 'should generate base64 secret' do
      Secret::generate_secret({'method'=>'base64','bytes'=>10}).length.should eq(16)
      res = Secret::generate_secret({'method'=>'base64','bytes'=>10000})
      @test_chars.(res,@base64).should == true
    end

    it 'should generate y64 secret' do
      Secret::generate_secret({'method'=>'y64','bytes'=>10}).length.should eq(16)
      res = Secret::generate_secret({'method'=>'y64','bytes'=>10000})
      @test_chars.(res,@y64).should == true
    end

    it 'should generate alphabet-based secret' do
      Secret::generate_secret({'method'=>'alphabet','bytes'=>11}).length.should eq(16)
      res = Secret::generate_secret({'method'=>'alphabet','bytes'=>10000})
      @test_chars.(res,@alphabet).should == true
    end

  end
end


test/scripts/setup-puppet-master.sh
echo "192.168.2.11    puppet-agent.local" >> /etc/hosts

cp /vagrant/fileserver.conf /etc/puppet/fileserver.conf

puppet master --mkusers --autosign true

mkdir -p /secrets
chown -R puppet:puppet /secrets


spec/functions/secret/secret_spec.rb
require 'spec_helper'
require 'rspec-puppet'
require 'puppet/file_serving/configuration'
require 'fileutils'

describe 'secret' do
  include RSpec::Puppet::Support

  # since let doesn't seem to work here
  #     let(:node) { 'spec.ops' }
  # instead pull in the fqdn manually:
  let(:scope) { PuppetlabsSpec::PuppetInternals.scope }

  it "should exist" do
    Puppet::Parser::Functions.function("secret").should == "function_secret"
  end

  describe "secret called by the client" do

    # make sure to bootstrap whatever environment we need
    before :each do
      @fqdn = 'spec.ops'
      @base_dir = "/tmp/secrets"
      FileUtils::rm_rf @base_dir

      # specify the fqdn
      scope.expects(:lookupvar).with("::fqdn").returns(@fqdn).at_least_once

      # mockup our configuration
      # mount to puppet:///secrets  ==>  /tmp/secrets/%H
      conf = Puppet::FileServing::Configuration.configuration
      mount = Puppet::FileServing::Mount::File.new "secrets"
      mount.path = "#{@base_dir}/%H"
      conf.stubs(:mounts).returns("secrets" => mount)
    end

    it 'should be callable without parameters' do
      scope.function_secret([]).should eq('puppet:///secrets/default')
    end

    it 'should define a secret with its first parameter' do
      scope.function_secret(["mysecret"]).should eq('puppet:///secrets/mysecret')
      scope.function_secret(["myother"]).should eq('puppet:///secrets/myother')
      File::exists?(File::join( @base_dir, @fqdn, 'mysecret')).should == true
      File::exists?(File::join( @base_dir, @fqdn, 'myother')).should == true
    end

    it 'should reuse the secret once it was created' do
      scope.function_secret(["one"]).should eq('puppet:///secrets/one')
      f1 = File::read File::join( @base_dir, @fqdn, 'one')

      scope.function_secret(["one"]).should eq('puppet:///secrets/one')
      f2 = File::read File::join( @base_dir, @fqdn, 'one')

      f1.should eq(f2)
    end

  end

end


lib/puppet/parser/functions/secret.rb

Puppet::Parser::Functions::newfunction(:secret, :type => :rvalue) do |vals|
  secretid = vals[0] || 'default'
  opts = vals[1] || {}
  opts['secrets_mount'] ||= 'secrets'

  # get the callee (secrets are saved based on fqdn)
  callee = lookupvar('::fqdn')
  raise Puppet::ParseError, "Can't find the FQDN of the client that is using secret(). aborting." if callee == :undefined

  Secret::ensure_secret callee, secretid, opts
  
  # point the client to his secret
  return "puppet:///#{opts['secrets_mount']}/#{secretid}"
end

require 'securerandom'
# Backport of missing SecureRandom methods from 1.9
# http://softover.com/UUID_in_Ruby_1.8
module SecureRandom
  class << self
    def method_missing(method_sym, *arguments, &block)
      case method_sym
      when :uuid
        r19_uuid(*arguments)
      else
        super
      end
    end

    private
    def r19_uuid
      ary = random_bytes(16).unpack("NnnnnN")
      ary[2] = (ary[2] & 0x0fff) | 0x4000
      ary[3] = (ary[3] & 0x3fff) | 0x8000
      "%08x-%04x-%04x-%04x-%04x%08x" % ary
    end
  end
end

module Secret
  class << self
    # global definitions, adjust to your liking
    MAX_SECRET_BYTES = 10*1024
    MIN_SECRET_BYTES = 1
    IDENTIFIER_ALPHABET =
      "abcdefghijklmnopqrstuvwxyz"+
      "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    def ensure_secret callee, secretid, opts = {}
      # validate all user input
      validate callee, 'FQDN'
      validate secretid, 'secret ID'
      # do the secret defining action here...
      secrets_dir = get_secrets_dir callee, opts["secrets_mount"] || "secrets"
      secret_file = File::join secrets_dir, secretid
      create_secret secret_file, opts if not File::exists? secret_file
    end

    def generate_secret opts = {}
      # how bytes in the secret
      method = opts['method'].to_s
      len = opts['length'].to_i
      if len > 0
        bytes = map_length_to_bytes len, method
      else
        bytes = ( opts['bytes'] || 128 ).to_i
      end

      # make sure we don't have too few or too many bytes
      if    bytes < MIN_SECRET_BYTES
        raise Puppet::ParseError, "secrets cannot have a length of less than #{MIN_SECRET_BYTES} bytes (you provided '#{opts['bytes']}'). aborting."
      elsif bytes > MAX_SECRET_BYTES
        raise Puppet::ParseError, "secrets cannot have a length of more than #{MAX_SECRET_BYTES} bytes (you provided '#{opts['bytes']}'). aborting."
      end

      generate_secret_for_method method, bytes, opts
    end


    private

    def validate sth, name
      if sth.to_s.empty?
        raise Puppet::ParseError, "#{name} for callee of function 'secret()' is empty. can't proceed without it."
      end

      if (sth =~ /^[^\/]*$/).nil?
        raise Puppet::ParseError, "#{name} for callee of function 'secret()' contains '/' (you provided '#{sth}'). this is not allowed."
      end

      if not (sth =~ /^[.]+$/).nil?
        raise Puppet::ParseError, "#{name} for callee of function 'secret()' contains only dots ('.') (you provided '#{sth}'). this is not allowed."
      end
    end

    def create_secret secret_file, opts = {}
      secret = generate_secret opts
      write_secret_to_file secret, secret_file
    end

    def alphabet_secret bytes, alphabet
      raise Puppet::ParseError, "trying to encode with an empty alphabet. this is impossible." if alphabet.empty?

      res = ''
      n = alphabet.length
      length = bytes_to_length_for_cardinality bytes, n
      for i in (0..(length-1))
        res += alphabet[ SecureRandom.random_number(n), 1 ]
      end
      res
    end

    def y64_secret bytes
      SecureRandom.base64(bytes).gsub('+','.').gsub('/','_').gsub('=','-')
    end

    def ceph_secret
      if which("ceph-authtool").nil?
        raise Puppet::ParseError, "you must install ceph and have ceph-authtool executable in order to generate ceph secrets. aborting."
      else
        `ceph-authtool --gen-print-key`
      end
    end

    def length_to_bytes_for_cardinality length, cardinality
      ( length * (Math::log(cardinality)/Math::log(2)) / 8 ).floor
    end

    def bytes_to_length_for_cardinality bytes, cardinality
      ( bytes * 8 / (Math::log(cardinality)/Math::log(2)) ).ceil
    end

    def map_length_to_bytes length, method
      case method
      when 'base64'     ; length_to_bytes_for_cardinality length, 64
      when 'y64'        ; length_to_bytes_for_cardinality length, 64
      when 'alphabet'   ; length_to_bytes_for_cardinality length, IDENTIFIER_ALPHABET.length
      # the default has bytes == length; all methods that don't care about length can be safely ignored
      else              ; length
      end
    end

    def generate_secret_for_method method, bytes, opts = {}
      case method
      when 'base64'     ; SecureRandom.base64 bytes
      when 'y64'        ; y64_secret bytes
      when 'alphabet'   ; alphabet_secret bytes, IDENTIFIER_ALPHABET
      when 'default','' ; SecureRandom.random_bytes bytes
      when 'ceph'       ; ceph_secret
      when 'uuid'       ; SecureRandom.uuid
      else                raise Puppet::ParseError, "don't understand method '#{opts['method']}' for secret generation. aborting."
      end
    end


    def write_secret_to_file secret, secret_file
      begin
        f = File.new secret_file, 'w'
        f.puts secret
        f.close
      rescue Errno::EACCES
        raise Puppet::ParseError,
            "could not write to secret file '#{secret_file}'. "+
            "check your permissions and make sure puppet has write/create access to the location. "+
            "can't generate secret without a secret folder."
      end
    end

    def get_secrets_dir node, mount_point
      require 'puppet/util/inifile'
      require 'fileutils'

      fileserver = Puppet::FileServing::Configuration.configuration
      dir_obj = fileserver.find_mount mount_point, nil

      # make sure the mountpoint actually exists
      if dir_obj.nil?
        raise Puppet::ParseError,
          "can't find a folder mounted to '#{mount_point}'. "+
          "make sure you have a folder configured in fileserving for mountpoint '#{mount_point}'. "+
          "can't generate secret without a secret folder."
      end

      # get the specific folder for this callee
      dir = dir_obj.path node

      # make sure the secret folder exists
      if not File::exists? dir
        begin
          FileUtils::mkdir_p dir
        rescue Errno::EACCES
          raise Puppet::ParseError,
            "could not create secret folder '#{dir}' for mount point '#{mount_point}'. "+
            "check your permissions and make sure puppet has write/create access to the location. "+
            "can't generate secret without a secret folder."
        end
      end

      # make sure it is a folder
      if not File::directory? dir
        raise Puppet::ParseError,
          "secret path for mount point '#{mount_point}' is not a directory (#{dir}). "+
          "can't generate secret without a secret folder."
      end

      dir
    end

    # determine if a command exists:
    # http://stackoverflow.com/questions/2108727/which-in-ruby-checking-if-program-exists-in-path-from-ruby
    def which cmd
      exts = ENV['PATHEXT'] ? ENV['PATHEXT'].split(';') : ['']
      ENV['PATH'].split(File::PATH_SEPARATOR).each do |path|
        exts.each { |ext|
          exe = File.join(path, "#{cmd}#{ext}")
          return exe if File.executable? exe
        }
      end
      return nil
    end

  end
end